### PR TITLE
EigenSolversApp/Perturb_Geometry_Process

### PR DESCRIPTION
**Pull request by manuelmessmer**
This PR adds a geometry perturbation process to the EigenSolversApplication. With this process the user can perturb the initial geometry of a structure. This is especially useful in combination with the buckling solver in Kratos, as you can analyze
the buckling behaviour of imperfect geometries. 
The process is based on random field theory. In a nutshell: A correlation matrix (size:NxN) for all nodes in the model part is assembled. The correlation value is based on the euclidean distance between the nodes. If nodes are close together -> large value, if further apart -> small value. The correlation matrix is decomposed and the eigenvectors are superposed to form the random field. Each eigenvector has a random number associated to it, such that an arbitrary amount of unique random fields can be derived from the same eigenvectors. Because an eigensolver is required I added this process to the EigenSolversApplication.

The so-called correlation length is a user defined parameter, which roughly defines the characteristic wavelength of the final perturbations -> see images below for small and large correlation length values. In both images the original geometry is a straight plate.
In theory the correlation matrix is an NxN dense matrix. To be able to apply this process also to larger models, there are two different implementations. The first one assembles an artificially sparse correlation matrix and the second approach solves the random field problem in a subgrid (coarse mesh) and then maps the final perturbation field back to the full domain.

large_correlation_length
small_correlation_length

### CODE REVIEW

**Comment by philbucher** [comment]
I have some comments please wait for my review

**Comment by manuelmessmer** [comment]
@qaumann the failure of the CI here is somehow related to the FEAST eigensolver support. I also tested it on my machine and when FEAST is enabled it also crashes. However, even if I exclude all my changes in the compilation it still does not compile. It seems to fail due to the feast_eigensystem_solver.h.  Do you have any idea why this is happening?

**Comment by qaumann** [comment]
@manuelmessmer Are you using gcc10 on your machine? You need to add `-std=legacy` to the Fortran compiler arguments, I prepared a fix for that but was waiting for the application to be renamed. 

The issue in the CI is due to the redefinition of some FEAST functions in MKL. I excluded these definitions in `add_custom_solvers_to_python.h` but apparently, this is not the right place...

**Comment by manuelmessmer** [comment]
@qaumann no I am using an older version. Does it only work with gcc10?

Ok, but that means I have to wait with this PR until you merged your fixes anyway, right?

**Comment by qaumann** [comment]
> 
> 
> @qaumann the failure of the CI here is somehow related to the FEAST eigensolver support. I also tested it on my machine and when FEAST is enabled it also crashes. However, even if I exclude all my changes in the compilation it still does not compile. It seems to fail due to the feast_eigensystem_solver.h. Do you have any idea why this is happening?

@manuelmessmer Are you using gcc10 on your machine? You need to add `-std=legacy` to the Fortran compiler arguments, I prepared a fix for that but was waiting for the application to be renamed. 

The issue in the CI is due to the redefinition of some FEAST functions in MKL. I excluded these definitions in `add_custom_solvers_to_python.h` but apparently, the MKL include happens before that. Strange though that it worked for the last pull requests.

**Comment by qaumann** [comment]
> 
> 
> @qaumann no I am using an older version. Does it only work with gcc10?
> 
> Ok, but that means I have to wait with this PR until you merged your fixes anyway, right?

Ok interesting, I will have a look. But the problem here is not the Fortran compiler, it is the redefintion of MKL's FEAST implementation that should be fixed by the `#define` in `add_custom_solvers_to_python.h`. I will check that.

**Comment by qaumann** [comment]
I just compiled this branch on my machine without any errors, perhaps the error is related to cotire?

@philbucher The CI runs for the other branches, right?

**Comment by oberbichler** [comment]
Good work but I don't like that this process becomes part of the SolversApplication. Is there no possibility to use the solver from another application?

File: applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp
Comment (philbucher) [inline applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: why don't you add this like the `EigensystemSolver`? 
Then eventually you could use it through the `EigenSolversFactory` (Eigen NOT referring to the library)

**Review by philbucher — CHANGES_REQUESTED** [review]
> Good work but I don't like that this process becomes part of the SolversApplication. Is there no possibility to use the solver from another application?

I fully agree with @oberbichler 
The `EigenSolversApp` is being renamed to `LinearSolversApp` in #6888 and becomes the main app for linear solvers in Kratos

Unfortunately the proposed additions don't fit there, for me those could go into the StructuralMechanicsApp

Looking at the code, you are using the Eigen-matrices for the computations, why? Until now our "official" linear algebra package is Ublas and even though we were discussing Eigen as an alternative it is very likely we will never switch to it
=> I don't like supporting internally a second package

Can you please elaborate why you are using Eigen? (except that it is probably more convenient)
I would like to understand more your motivation for going this way before recommending how to proceed further

BTW I recommend to discuss such potentially "political" changes before you start to implement, this way (hopefully) there will be less overhead :)

File: applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp
Comment (manuelmessmer) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: I used the Eigen-matices in this file, because of the function argument of the eigensolver. But yes you are right, I will add it to the EigenSolversFactory and then I can wrap the Ublas matrix there..
Nevertheless, I can still not use the eigensolvers inside the StructuralMechanicsApp right? Or is there any way to do so?

**Comment by manuelmessmer** [comment]
> BTW I recommend to discuss such potentially "political" changes before you start to implement, this way (hopefully) there will be less overhead :)

Will do ;)

File: applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp
Comment (philbucher) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: > I used the Eigen-matices in this file, because of the function argument of the eigensolver. But yes you are right, I will add it to the EigenSolversFactory and then I can wrap the Ublas matrix there..

sounds good

@loumalouomega there is no C++ EigenSolverFactory yet right? I guess this is waiting for #3185 ?

> Nevertheless, I can still not use the eigensolvers inside the StructuralMechanicsApp right? Or is there any way to do so?

you can if you use the factory (which is currently only available in Python, so you'd have to make an interface for it, just like e.g. the B&S has)

Comment (manuelmessmer) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: Understand. Makes sense. Then for now we agree that I move everything to the StructuralMechanicsApp with a python interface?
Shall I open a new PR or do the changes here?

Comment (philbucher) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: I guess a new PR would be cleaner :)

Comment (philbucher) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: > Then for now we agree that I move everything to the StructuralMechanicsApp with a python interface?

yes, I guess this is the best place. Or do you have another suggestion?

Comment (manuelmessmer) [reply applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp:107]: No, totally agree. Will work on that..

**Comment by manuelmessmer** [comment]
I was just trying to add a "DenseEigenvalueSolver" to the EigenSolverFactory. The Problem is that the layout for the Solve function would look like this: LinearSolver.Solve( DenseMatrixType rA, DenseVectorType rEigenvalues, DenseMatrixType rEigenvectors). However, this argument list is not supported by the base class. Is there any better way than adding this argument list to the LinearSolver base class?
I was also thinking about doing something like: Solve( SparseMatrixType rA, SparseMatrixType rDummy, DenseVectorType rEigenvalue, DenseMatrixType rEigenvectors ) and then wrap rA into a Dense Eigen::MatrixXd inside the function.
Do you have any suggestions?

**Comment by oberbichler** [comment]
It is not a good idea to implement an eigenvalue solver as linear solver (Except as a temporary hack).
For a clean solution an interface for an eigenvalue solver would have to be written. As far as I understand this is the reason why you have integrated this process directly into the EigenSolversApplicaiton(?)

**Comment by manuelmessmer** [comment]
Well, yes that was the problem. Therefore, I wanted to derive the eigenvalue solver from the linear_solver.h and implement a similar interface as it was done for the eigensystem_solver and FEAST_eigensystem_solver (which are also derived from linear_solver.h).  To suit the aforementioned eigensystem_solvers the linear_solver.h has a unique implementation of the Solve()-function. That's why I thought about having another one for a standard eigenvalue solver...

Do you mean to add an eigenvalue solver base class to the core?

**Comment by philbucher** [comment]
@manuelmessmer are you at the chair one of these days?
Probably easier to discuss in person

**Comment by oberbichler** [comment]
> a similar interface as it was done for the eigensystem_solver

Now I remember... Well I think we can keep this workaround at the moment 😄

**Comment by manuelmessmer** [comment]
> Ok interesting, I will have a look. But the problem here is not the Fortran compiler, it is the redefintion of MKL's FEAST implementation that should be fixed by the `#define` in `add_custom_solvers_to_python.h`. I will check that.

@qaumann, in the mkl version I have installed, the '_MKL_SOLVERS_EE_H' in 'mkl/include/mkl_solvers_ee.h' is defined with two underscores in the beginning. (__MKL_SOLVERS_EE_H). That was the reason, why it did not compile on my machine. Just wanted to let you know, in case someone will encounter the same problem.

**Comment by philbucher** [comment]
this can be closed now right?

**Pull request closed without merging**
